addons/Bonanza/Bonanza.lua

```diff
@@ -0,0 +1,195 @@
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL from20020516 BE LIABLE FOR ANY
+DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.]]
+
+_addon.name = 'Bonanza'
+_addon.author = 'from20020516'
+_addon.version = '1.1'
+_addon.command = 'bonanza'
+
+extdata = require('extdata')
+packets = require('packets')
+require('logger')
+
+marble = {id=2559,price=2000,limit=10} --general settings
+
+win_numbers = {}
+math.randomseed(os.time())
+
+--search your bags for bonanza marble and get number then add prefix 0
+function search_marbles()
+  local marbles = {}
+  local bags = {0,1,2,4,5,6,7,9}
+  for bag_ind=1,#bags do
+    for index=1,windower.ffxi.get_bag_info(bags[bag_ind]).max do
+      local item = windower.ffxi.get_items(bags[bag_ind],index)
+      if item.id == marble.id then --bonanza marble
+        local five_digits = string.format("%05s",extdata.decode(item).number)
+        table.insert(marbles,{bags[bag_ind],index,five_digits})
+      end
+    end
+  end
+  return marbles;
+end
+
+windower.register_event('addon command',function(...)
+  moogle = windower.ffxi.get_mob_by_name('Bonanza Moogle')
+  decides = {}
+  local cmd={...}
+  local count = marble.limit-#search_marbles()
+
+  for i=1,#cmd do
+    math.random() --shuffle rand
+    cmd[i] = tonumber(cmd[i]) and cmd[i]*1 or cmd[i]
+  end
+
+  if cmd[1] == 'judge' then
+    windower.send_command('input /smes')
+  elseif count == 0 then
+    error('you have already '..marble.limit..' marbles.')
+  elseif cmd[1] == 'random' then
+    for i=1,count do
+      table.insert(decides,math.random(0,99999))
+    end
+  elseif cmd[1] == 'sequence' then
+    local n = math.min(cmd[2],10000-count)
+    for i=n,n+count-1 do
+      table.insert(decides,i)
+    end
+  elseif cmd[1] == 'last' and 10 > cmd[2] then
+    for i=1,count do
+      table.insert(decides,tonumber(math.random(0,9999)..cmd[2]))
+    end
+  elseif 100000 > cmd[2] and not cmd[marble.limit+1] then
+    decides = cmd
+  end
+
+  if #decides > 0 then
+    talk_moogle()
+  end
+end)
+
+function talk_moogle()
+  --print(unpack(decides))
+  if #decides > 0 then
+    hide_ui = true
+    local packet = packets.new('outgoing',0x01A,{
+      ["Target"]=moogle.id,
+      ["Target Index"]=moogle.index,
+      ["Category"]=0,
+      ["Param"]=0,
+      ["_unknown1"]=0})
+    packets.inject(packet)
+  end
+end
+
+windower.register_event('incoming chunk',function(id,data)
+  --got a marble.
+  if id == 0x020 and hide_ui and data:unpack('H',0x0D) == marble.id and data:unpack('C',0x0F) == 0 then
```

Comment:
You can use the `packets` library here as well to not manually unpack the data:
```lua
    local packet = packets.parse('incoming', data)
    if packet.id == 0x020 and hide_ui and packet['Item'] == marble.id and packet['Bag'] == 0 then
        -- And so on
```

That said, your numbers look off to me. `0x0D` is the high order byte of the item ID, if I'm reading this right. That should be `0x0C`. And `0x0F` is the inventory index... you probably meant `0x0E` here? I would double check these numbers if I were you.

Reply:
umm, but it works. It's probably because the .lua table starts with 1. (not 0)

Reply:
You're right, that explains it :)